### PR TITLE
Add example for OpenTelemetry Tracer

### PR DESCRIPTION
**Pull request by AlexanderEllis**
Commit Message: Add example for OpenTelemetry Tracer
Additional Description: This is the Jaeger Tracing demo, but with OpenTelemetry tracing instead. Like the Jaeger Tracing demo, it pulls up 3 Envoy instances with OTLP tracing enabled, then also runs the OpenTelemetry Collector.
Risk Level: Very low (just an example)
Testing: Added verify.sh
Docs Changes: Added README.md here, but happy to add docs as well
Part of https://github.com/envoyproxy/envoy/issues/9958

### CODE REVIEW

**Comment by phlax** [comment]
hi @AlexanderEllis thans for working on this - very useful - headsup that these sandboxes can take quite a bit of back and forth

we will definitely need some rst docs for this - im happy to help putting these together

i see this is WIP still - do you want me to give it a first pass now ?

**Comment by AlexanderEllis** [comment]
Hi @phlax , sounds good to me! Last thing I'm fixing is just ensuring the verify script passes in the pipeline as well — I think I was running into permissions issues with the yaml config for the collector, but I've added a bit to (hopefully) fix that part.

A first pass would be great! I'll mark this as review, and happy to iterate as we go :) Thanks!

**Comment by AlexanderEllis** [comment]
@phlax Still running into trouble with the verify script in the [CI pipeline](https://dev.azure.com/cncf/envoy/_build/results?buildId=115222&view=logs&jobId=f1a49366-51c0-540f-0294-f73c2bb311a9&j=f1a49366-51c0-540f-0294-f73c2bb311a9&t=7fd036c9-61ec-5b59-3645-e90950dc67a6), though things are working locally 🤔  I'm currently seeing the following:

```
opentelemetry_collector_1  | Error: failed to get config: cannot resolve the configuration: unable to read the file file:/etc/otel-collector-config.yaml: open /etc/otel-collector-config.yaml: permission denied
opentelemetry_collector_1  | 2022/08/18 20:03:29 collector server run finished with error: failed to get config: cannot resolve the configuration: unable to read the file file:/etc/otel-collector-config.yaml: open /etc/otel-collector-config.yaml: permission denied
```

I _think_ I'm adding this file correctly in the Dockerfile:

```
  opentelemetry_collector:
    build:
      context: .
      dockerfile: Dockerfile-opentelemetry-collector
    command: ["--config=/etc/otel-collector-config.yaml"]
    volumes:
    - ./otel-collector-config.yaml:/etc/otel-collector-config.yaml
    ports:
    - "${PORT_COLLECTOR_ZPAGE:-55679}:55679"
```

Any suggestions on this? I'm far from a docker expert, and any help would be appreciated.

**Comment by phlax** [comment]
> Any suggestions on this? I'm far from a docker expert, and any help would be appreciated.

i will review tomorrow - but off hand my guess is that the permissions in CI are different.

before running it the permissions/umask are restricted - in ci/do_ci.sh

```bash
 191   umask 027
 192   chmod -R o-rwx examples/
```

it does this to simulate an env (ie google) where home dirs are shared and perms are stricter

File: examples/opentelemetry-tracing/Dockerfile-opentelemetry-collector [hunk 1]
```diff
@@ -0,0 +1 @@
+FROM otel/opentelemetry-collector:latest
```
Comment (phlax) [inline examples/opentelemetry-tracing/Dockerfile-opentelemetry-collector:1]: can you add the `@sha:xyz` here as then its pinned, and will get updated by dependabot

File: examples/opentelemetry-tracing/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      replicas: 0
+
+  # Front envoy
+  front-envoy:
+    build:
+      context: .
+      dockerfile: Dockerfile-frontenvoy
+    ports:
+    - "${PORT_PROXY:-8000}:8000"
+    - "${PORT_PROXY_ADMIN:-8001}:8001"
```
Comment (phlax) [inline examples/opentelemetry-tracing/docker-compose.yaml:28]:
```suggestion
    - "${PORT_ADMIN:-8001}:8001"
```

File: examples/opentelemetry-tracing/service2-envoy-opentelemetry.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    load_assignment:
+      cluster_name: opentelemetry_collector
+      endpoints:
+      - lb_endpoints:
+        - endpoint:
+            address:
+              socket_address:
+                address: opentelemetry_collector
+                port_value: 4317
+admin:
```
Comment (phlax) [inline examples/opentelemetry-tracing/service2-envoy-opentelemetry.yaml:72]: if we include admin, then we should add some docs (and verify.sh test) showing how it is affected by calls to/through the proxy

File: examples/opentelemetry-tracing/service1-envoy-opentelemetry.yaml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    load_assignment:
+      cluster_name: opentelemetry_collector
+      endpoints:
+      - lb_endpoints:
+        - endpoint:
+            address:
+              socket_address:
+                address: opentelemetry_collector
+                port_value: 4317
+admin:
```
Comment (phlax) [inline examples/opentelemetry-tracing/service1-envoy-opentelemetry.yaml:123]: same here wrt to admin in the config

File: examples/opentelemetry-tracing/service2-envoy-opentelemetry.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
+static_resources:
```
Comment (phlax) [inline examples/opentelemetry-tracing/service2-envoy-opentelemetry.yaml:1]: its not immediately clear to me why we have 2 envoy configs

File: examples/opentelemetry-tracing/docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+version: "3.8"
+services:
+
+  # base images
+  flask_service:
```
Comment (phlax) [inline examples/opentelemetry-tracing/docker-compose.yaml:5]: i think this is probably necessary - as the final destination for any requests, but i dont see it used

File: examples/opentelemetry-tracing/verify.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash -e
+
+export NAME=opentelemetry-tracing
+export PORT_PROXY="${PORT_PROXY:-8000}"
+export PORT_COLLECTOR_ZPAGE="${PORT_COLLECTOR_ZPAGE:-55679}"
```
Comment (phlax) [inline examples/opentelemetry-tracing/verify.sh:5]: the reason for this pattern is to allow the examples to be tested simultaneously (ie to expose different ports when run with verify.sh)

once its clearer what services/ports need to be exposed i can suggest what ports to use here

**Review by phlax — COMMENTED** [review]
ive added a few high level comments - there were a few others that i held back on as i wasnt totally clear how the example is expected to work

im imagining that it should work something like

user --> proxy --> backend-service
                 |
                \/
             collector

the user then makes a call to the backend service via the proxy, and then looks at the collector to see any traces

perhaps this is not quite right, but would be good to clarify

i didnt review the docs yet - but they need to be moved to rst and crucially the steps in the docs _must_ ~match the steps (ie `run_log` steps) in verify.sh

one thing to emphasize is that these examples are foremost targetted at end users, the verify.sh script, while useful as an integration test, is meant to be validation that the steps described work (and continue to do so)

**Comment by phlax** [comment]
following up on this - im just looking at flakes in other examples

the shared/tracing Docker image is probably best avoided - its really not an example of best practice as it has 2 processes running inside the container

im just looking at how we can unpick this from existing examples

**Comment by phlax** [comment]
im realizing that this PR is copying out some of the issues in other tracing examples

it was on my agenda to rewrite these - esp for zipkin as it can flake - but also for the jaeger ones

im most of the way through rewriting the zipkin example - which can be seen here #22771, altho its going to take me a while to complete

**Comment by phlax** [comment]
zipkin sandbox update is complete - i would use that (ie in PR #22771 as it hasnt yet landed) as the basis for this sandbox

updated docs for the zipkin sandbox can be seen here https://storage.googleapis.com/envoy-pr/87b2474/docs/start/sandboxes/zipkin_tracing.html

**Comment by moderation** [comment]
With OpenTelemetry now being supported (thanks @AlexanderEllis!), I think we can / should deprecate the Jaeger and OpenTracing extensions and sandboxes. Discussion at https://github.com/envoyproxy/envoy/issues/9958. @cgilmour is planning to update the Datadog tracer to remove the OpenTracing dependency which will allow for its removal - https://github.com/envoyproxy/envoy/issues/21083#issuecomment-1196172567

**Comment by phlax** [comment]
> I think we can / should deprecate the Jaeger and OpenTracing extensions and sandboxes

it would be good to clarify which sandboxes these are - there are a few other tracing sandboxes (eg skywalking, zipkin)

either way - i think the pattern that the updated zipkin sandbox uses is roughly the pattern we want for the tracing sandboxes

**Comment by phlax** [comment]
@AlexanderEllis was hoping to ping you on slack - but didnt find you there - so i guess ill post here

@nguyenptk has been working on a Kafka sandbox and has got a pretty good idea how the sandbox docs/verify work

im wondering if it would be helpful for him and I to take over this PR - im thinking we can use the settings you have set up here but with the zipkin example as a template

I really appreciate the work you have done/are doing on the OT extension - so happy to work on this if it is helpful

**Comment by AlexanderEllis** [comment]
Hey @phlax , sorry for the delay here (day-job got busy), but that would be great! It sounds like the new Zipkin example would be great to reuse here, and from the sounds of it, the only differences would be 1) the OpenTelemetry Tracer config 2) seeing the traces and 3) viewing the OpenTelemetry collector page.

Happy to help with any questions about the setup 👍  Thanks again!

**Comment by phlax** [comment]
closing this as #22988 has now landed

**Pull request closed without merging**